src/target/adiv5.c

```diff
@@ -822,16 +831,22 @@ void adiv5_dp_init(adiv5_debug_port_s *const dp, const uint32_t idcode)
 			(designer & ADIV5_DP_DESIGNER_JEP106_CONT_MASK) << 1U | (designer & ADIV5_DP_DESIGNER_JEP106_CODE_MASK);
 		dp->partno = (dpidr & ADIV5_DP_DPIDR_PARTNO_MASK) >> ADIV5_DP_DPIDR_PARTNO_OFFSET;
 
+		/* Minimal Debug Port (MINDP) functions implemented */
 		dp->mindp = !!(dpidr & ADIV5_DP_DPIDR_MINDP);
 
-		/* Check for a valid DPIDR / designer */
-		if (dp->designer_code != 0) {
+		/* 
+		 * Check DPIDR validity
+		 * Designer code 0 is not a valid JEP-106 code
+		 * Version 0 is reserved for DPv0 which does not implement DPIDR
+		 * Bit 0 of DPIDR is read as 1
+		 */
+		if (dp->designer_code != 0 && dp->version > 0 && dpidr & 1U) {
```

Comment:
The masking of `dipid` here needs parens to get the right order of operations:
```suggestion
		if (dp->designer_code != 0 && dp->version > 0 && (dpidr & 1U)) {
```

Without this, it's equivalent to having written:
```c
		if ((dp->designer_code != 0 && dp->version > 0 && dpidr) & 1U) {
```

Reply:
Actually, the Bitwise AND should take precedence over Logical AND, at least according to the [cheat sheet](https://en.cppreference.com/w/c/language/operator_precedence)

Do let me know if I'm reading this wrong or using the wrong resource tho.

Reply:
Ah, you're right it seems. This is the problem with & - if you're doing, eg, `(dpidr & 1U) != 0` then the parens are absolutely required.. the confusing precedence order is as a result a lint on many compilers.

Looks like technically what you wrote is correct then, so we'll put the request this way: For clarity so this doesn't become a constant query, how about we include the parens in this case so the reader doesn't make the same mistake we did, even though they're not strictly required? We suspect we'll forget this order of operations clarification 6mo down the line, so making it more explicit can only be a good thing in context.

Reply:
Sounds perfectly reasonable.